File: 2022/day07/main.ts
#!/usr/bin/env deno run --allow-read
import { assertEquals } from "https://deno.land/std@0.116.0/testing/asserts.ts";
import { matchGroups, sum } from "../../2020/utils.ts";

const parseInput = (string: string): [string, string[]][] =>
  string.trim().split("$ ").filter((b) => b).map((block) => {
    const [command, ...outputLines] = block.trim().split("\n");
    return [command, outputLines];
  });

const text = await Deno.readTextFile("input.txt");

const input = parseInput(text);

type Directory = {
  name: string;
  size: number;
  totalSize: number;
  children: Directory[];
};

const dirMatcher = matchGroups(/^(?<sizeString>\d+) (?<filename>.+)$/);

const reconstructDirectories = (input: [string, string[]][]): Directory => {
  const root = { name: "", size: 0, totalSize: -1, children: [] };
  const currentDirs: Directory[] = [];

  for (const [command, lines] of input) {
    const [bin, ...args] = command.split(" ");
    if (bin === "cd") {
      if (lines.length) {
        throw new Error(`Can not handle lines for cd: ${lines}`);
      }
      const [path] = args;
      if (path === "..") currentDirs.pop();
      else if (path === "/") {
        while (currentDirs.length) currentDirs.pop();
        currentDirs.push(root);
      } else {
        let child = currentDirs.at(-1)?.children.find((c) => c.name === path);
        if (!child) {
          child = { name: path, size: 0, totalSize: -1, children: [] };
          currentDirs.at(-1)?.children.push(child);
        }
        currentDirs.push(child);
      }
    } else if (bin === "ls") {
      if (args.length) {
        throw new Error(`Can not handle args for ls: ${args}`);
      }
      const currentDir = currentDirs.at(-1);
      if (!currentDir) throw new Error("Running ls before a cd");
      currentDir.size = 0;
      for (const line of lines) {
        if (line.startsWith("dir ")) continue;
        const { sizeString } = dirMatcher(line);
        currentDir.size += parseInt(sizeString, 10);
      }
    } else {
      throw new Error(`Unknonw command: ${command}`);
    }
  }

  const figureTotalSize = (dir: Directory): number => {
    return (dir.totalSize = dir.size + sum(dir.children.map(figureTotalSize)));
  };

  figureTotalSize(root);

  return root;
};

const part1 = (input: [string, string[]][]): number => {
  const root = reconstructDirectories(input);

  const sumSuitableTotalSizes = (dir: Directory): number => {
    const mine = dir.totalSize <= 100000 ? dir.totalSize : 0;
    const others = sum(dir.children.map(sumSuitableTotalSizes));
    return mine + others;
  };

  return sumSuitableTotalSizes(root);
};

const exampleText = await Deno.readTextFile("example.txt");

const example = parseInput(exampleText);

assertEquals(part1(example), 95437);

console.log("Result part 1: " + part1(input));

const part2 = (input: [string, string[]][]): number => {
  const totalSpace = 70000000;
  const requiredFreeSpace = 30000000;

  const root = reconstructDirectories(input);

  const currentFreeSpace = totalSpace - root.totalSize;
  const minToDelete = requiredFreeSpace - currentFreeSpace;
  if (minToDelete < 0) {
    throw new Error("Already enough space!");
  }

  const candidateSizes: number[] = [];

  const findCandidateSizes = (dir: Directory): void => {
    if (dir.totalSize >= minToDelete) {
      candidateSizes.push(dir.totalSize);
    }
    dir.children.forEach(findCandidateSizes);
  };

  findCandidateSizes(root);

  return Math.min(...candidateSizes);
};

assertEquals(part2(example), 24933642);

console.log("Result part 2: " + part2(input));

File: 2022/day07/main.rs
use std::collections::HashSet;
use std::fmt::Error;
use std::rc::Rc;

#[derive(Debug)]
struct CommandOutput {
    command: String,
    lines: Vec<String>,
}

impl std::str::FromStr for CommandOutput {
    type Err = Error;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        let (a, b) = input.split_once("\n").unwrap();

        Ok(Self {
            command: a.to_owned(),
            lines: b.lines().map(|s| s.to_owned()).collect(),
        })
    }
}

fn parse_input(input: &str) -> Vec<CommandOutput> {
    input
        .trim()
        .split("$ ")
        .filter(|block| block != &"")
        .map(|block| block.parse().unwrap())
        .collect::<Vec<_>>()
}

#[derive(Debug)]
struct Directory {
    name: String,
    size: u32,
    children: Vec<Rc<Directory>>,
}

fn reconstruct_directories(input: &Vec<CommandOutput>) -> Directory {
    // TODO: figure out ownership

    let mut root = Rc::new(Directory {
        name: "".to_owned(),
        size: 0,
        children: vec![],
    });
    let mut current_directories = vec![&mut root];

    for command_output in input {
        let command_bits = command_output.command.split(" ").collect::<Vec<_>>();
        match command_bits[..] {
            ["cd", path] => match path {
                ".." => {
                    current_directories.pop();
                }
                "/" => {
                    // current_directories = vec![&mut root];
                }
                name => {
                    let mut newDir =  Rc::new(Directory {
                        name: name.to_owned(),
                        size: 0,
                        children: vec![],
                    });
                    current_directories.last_mut().unwrap().children.push(newDir);
                    current_directories.push(Rc::clone(&newDir));
                }
            },
            ["ls"] => {}
            _ => {
                panic!("unknown command {}", command_output.command)
            }
        }
    }

    root
}

fn part1(input: &Vec<CommandOutput>) -> u32 {
    println!("dirs: {:?}", reconstruct_directories(input));

    95437
}

#[test]
fn check_part1() {
    assert_eq!(
        part1(&parse_input(
            &std::fs::read_to_string("day07/example.txt").unwrap()
        )),
        95437
    );
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = std::fs::read_to_string("day07/input.txt")?;

    let parsed_input = parse_input(&file);

    let part1 = part1(&parsed_input);
    println!("part 1: {}", part1);

    Ok(())
}
